### PR TITLE
Responsive address form that changes by country

### PR DESCRIPTION
**Pull request by sbinlondon**
Requires https://github.com/opencollective/opencollective-api/pull/5385
Resolves https://github.com/opencollective/opencollective/issues/3630

Uses the [Shopify address package and API](https://github.com/Shopify/quilt/tree/master/packages/address) to render a form that is responsive to the chosen country's postal format, including making fields required vs optional. Can also be localised based on what language the user chooses to browse Open Collective in.

### German address
Screenshot 2021-01-29 at 17 57 25
Screenshot 2021-01-29 at 17 56 52

### Japanese address
Screenshot 2021-01-29 at 18 16 15
Screenshot 2021-01-29 at 18 16 30

Not in scope of this PR: making sure the user address and payout method address match https://opencollective.slack.com/archives/G46PNRCGP/p1612218589031700 (https://github.com/opencollective/opencollective/issues/3340)

### CODE REVIEW

**Comment by sbinlondon** [comment]
Notes for further work on the PR:

* need double check that all the ISO codes from the InputCountryCode (and also the language choice dropdown) match the ones the Shopify API is expecting
* add a drop down select for province/state/etc based on the fields that are returned there

File: components/I18nAddressFields.js [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+/** Checks if the key from labels is also present in optionalLabels */
+const isFieldOptional = (addressInfo, fieldName) => {
+  return !Object.keys(addressInfo.optionalLabels).includes(fieldName);
+};
+
+/** Component to be used in forms that require addresses that need some validation
+ * i.e. Expenses and Contributions. Must be used with Formik. */
+const I18nAddressFields = ({ selectedCountry }) => {
+  const formik = useFormikContext();
```
Comment (Betree) [inline components/I18nAddressFields.js:75]: Not using Formik here would make things more flexible, and being able to use the component in non-formik forms would be a nice added value. What about using standard `onChange`/`value` props? Like `StyledSelect`, we would still be able to use it in a formik form like this:

```jsx
<Field name="location">
  {({ field }) => (
    <I18nAddressFields
      value={field.value}
      onChange={({ value, error }) => {
        if (error) {
          field.setError(error);
        } else {
          field.setValue(value);
        }
      }}
  )}
</Field>
```

**Review by Betree — COMMENTED** [review]
This component id going to be super helpful. A simple fallback if the query to shopify somehow fails could be to just display a simple textarea (like we do today). That would break the validations, but at least the forms where it's used would still work.

File: components/I18nAddressFields.js [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+      </StyledInputField>
+    );
+  }
+
+  // Zone select
+  const getZoneLabel = zone => {
+    return `${truncate(zone.name, { length: 30 })} - ${zone.code}`;
+  };
+
+  const getZoneSelectOptions = memoizeOne(zones => {
```
Comment (Betree) [inline components/I18nAddressFields.js:141]: Memoization here has no effect, `memoizeOne` will just return a new function for each render.

To make it work, you would have to wrap the callback passed to `memoizeOne` in `useCallback`.
But the simplest approach in functional components is to directly use `useMemo`.

File: components/I18nAddressFields.js [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    };
+
+    fetchData();
+  }, [selectedCountry]);
+
+  if (loading) {
+    return <LoadingPlaceholder width="100%" height={163} />;
+  }
+
+  // If the call to Shopify fails, fallback to the free text address field
```
Comment (Betree) [inline components/I18nAddressFields.js:117]: Nice :+1:

File: components/I18nAddressFields.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+/** Controlled component to be used in forms that require addresses that need some validation
+ * i.e. Expenses and Contributions. Can be used with or without Formik.
+ * See PropTypes info for differing props based on Formik vs. non-Formik.*/
+const I18nAddressFields = ({ selectedCountry, value, onChange }) => {
+  const intl = useIntl();
+
+  /** Pass user's chosen locale to AddressFormatter if present. */
+  if (intl.locale) {
+    addressFormatter.updateLocale(intl.locale);
+  }
```
Comment (Betree) [inline components/I18nAddressFields.js:85]: Would be great to wrap that in a `useEffect` that watches on `intl.locale`. Unless I'm missing something with the library, I think we only need to do this once on mount.

**Review by Betree — COMMENTED** [review]
Great! I haven't tested yet but this component's API and the comments used to document it are super clear :raised_hands: 

Some feedback for enhancements below.

**Comment by sbinlondon** [comment]
Have opened a [PR to address the expense form body rerendering](https://github.com/opencollective/opencollective-frontend/pull/5885) every time you take an action on the page.

This should help unblock displaying the errors. Sadly I could not figure it all out in time (would imperitavely set and display errors but then when the field was okay, it wouldn't clear the error) so hopefully it won't take too much more time next week to polish this off @Betree @kewitz

**Comment by kewitz** [comment]
Rebased.

**Pull request merged**